chains/orchestrator-relays/runtime/dancelight/src/bridge_to_ethereum_config.rs

```diff
@@ -345,6 +333,239 @@ where
     }
 }
 
+/// `NativeContainerTokensProcessor` is responsible for receiving and processing native container
+/// chain tokens coming from Ethereum and forwarding them to the container chain via Tanssi through XCM.
+pub struct NativeContainerTokensProcessor<
+    T,
+    EthereumLocation,
+    EthereumNetwork,
+    InboundQueuePalletInstance,
+>(
+    PhantomData<(
+        T,
+        EthereumLocation,
+        EthereumNetwork,
+        InboundQueuePalletInstance,
+    )>,
+);
+
+impl<T, EthereumLocation, EthereumNetwork, InboundQueuePalletInstance> MessageProcessor
+    for NativeContainerTokensProcessor<
+        T,
+        EthereumLocation,
+        EthereumNetwork,
+        InboundQueuePalletInstance,
+    >
+where
+    T: snowbridge_pallet_inbound_queue::Config
+        + pallet_ethereum_token_transfers::Config
+        + snowbridge_pallet_system::Config
+        + pallet_xcm::Config,
+    <T as frame_system::Config>::RuntimeEvent: From<RuntimeEvent>,
+    EthereumLocation: Get<Location>,
+    EthereumNetwork: Get<NetworkId>,
+    InboundQueuePalletInstance: Get<u8>,
+{
+    fn can_process_message(channel: &Channel, envelope: &Envelope) -> bool {
+        // Validate channel and gateway
+        if !GatewayAndChannelValidator::<T>::validate_gateway_and_channel(channel, envelope) {
+            log::warn!(
+                "NativeContainerTokensProcessor::can_process_message: invalid gateway or channel"
+            );
+            return false;
+        }
+
+        // Try decoding the message and check if the destination owns the token being transferred
+        if let Some(token_data) =
+            NativeTokenTransferData::decode_native_token_message(&envelope.payload)
```

Comment:
We do similar stuff in `can_process_message` and `process_message`.
Consider moving the logic to the common helper that will decode and extract the token location.